rllib/env/multi_agent_episode.py

```diff
@@ -0,0 +1,535 @@
+            agent_eps[agent_id].concat_episode(episode_chunk.agent_episodes[agent_id])
+            # Update our timestep mapping.
+            # TODO (simon): Check, if we have to cut off here as well.
+            self.global_t_to_local_t[agent_id][
+                :-1
+            ] += episode_chunk.global_t_to_local_t[agent_id]
+
+        self.t = episode_chunk.t
+        if episode_chunk.is_terminated:
+            self.is_terminated = True
+        if episode_chunk.is_truncated:
+            self.is_truncated = True
+
+        # Validate
+        # TODO (simon): Write validate function.
+        # self.validate()
+
+    # TODO (simon): Maybe adding agent axis. We might need only some agent observations.
+    # Then also add possibility to get __all__ obs (or None)
+    # Write many test cases (numbered obs).
+    def get_observations(self, indices: Union[int, List[int]] = -1, global_ts=True):
+        """Gets observations for all agents that stepped in the last timesteps.
+
+        Note that observations are only returned for agents that stepped
+        during the given index range.
+        """
+
+        # First for global_ts = True:
+        if global_ts:
+            # Check, if the indices are iterable.
+            if isinstance(indices, list):
+                indices = [self.t + idx if idx < 0 else idx for idx in indices]
+            else:
+                indices = [self.t + indices] if indices < 0 else [indices]
+
+            return {
+                agent_id: list(
+                    map(
+                        agent_eps.observations.__getitem__,
+                        self.global_t_to_local_t[agent_id].find_indices(indices),
+                    )
+                )
+                for agent_id, agent_eps in self.agent_episodes.items()
+                # Only include agent data for agents that stepped.
+                if len(self.global_t_to_local_t[agent_id].find_indices(indices)) > 0
+            }
+        # Otherwise just look for the timesteps in the `SingleAgentEpisode`s
+        # directly.
+        else:
+            # Check, if the indices are iterable.
+            if not isinstance(indices, list):
+                indices = [indices]
+
+            return {
+                agent_id: list(map(agent_eps.observations.__getitem__, indices))
+                for agent_id, agent_eps in self.agent_episodes.items()
+                # Only include agent data for agents that stepped so far at least once.
+                # TODO (sven, simon): This will not include initial observations. Should
+                # we?
+                if self.agent_episodes[agent_id].t > 0
+            }
+
+    def get_actions(self, indices: Union[int, List[int]] = -1, global_ts=True):
+        """Gets actions for all agents that stepped in the last timesteps.
+
+        Note that actions are only returned for agents that stepped
+        during the given index range.
+        """
+
+        # First for global_ts = True:
```

Comment:
Any chance we can unify this entire logic here in `get_observations|rewards|actions` into one or more helper utilities?

Reply:
I am positive. Give me some time to think about it

Reply:
Refactored into a single helper function `_getattr_by_index`